pubsub/kubemq/kubemq_events.go

```diff
@@ -0,0 +1,169 @@
+	return &kubeMQEvents{
+		client:               nil,
+		metadata:             nil,
+		logger:               logger,
+		publishFunc:          nil,
+		resultChan:           make(chan error, 1),
+		waitForResultTimeout: 60 * time.Second,
+		ctx:                  nil,
+		ctxCancel:            nil,
+	}
+}
+func (k *kubeMQEvents) init() error {
+	k.ctx, k.ctxCancel = context.WithCancel(context.Background())
+	if k.metadata.useMock {
+		k.isInitialized = true
+		return nil
+	}
+	clientID := k.metadata.clientID
+	if clientID == "" {
+		clientID = uuid.New()
+	}
+	client, err := kubemq.NewEventsClient(k.ctx,
+		kubemq.WithAddress(k.metadata.host, k.metadata.port),
+		kubemq.WithClientId(clientID),
+		kubemq.WithTransportType(kubemq.TransportTypeGRPC),
+		kubemq.WithCheckConnection(true),
+		kubemq.WithAuthToken(k.metadata.authToken),
+		kubemq.WithAutoReconnect(true),
+		kubemq.WithReconnectInterval(time.Second))
+	if err != nil {
+		k.logger.Errorf("error init kubemq client error: %s", err.Error())
+		return err
+	}
+	k.ctx, k.ctxCancel = context.WithCancel(context.Background())
+	k.client = client
+	if err := k.setPublishStream(); err != nil {
+		k.logger.Errorf("error init kubemq client error: %w", err.Error())
+		return err
+	}
+	k.isInitialized = true
+	return nil
+}
+func (k *kubeMQEvents) Init(meta *metadata) error {
+	k.metadata = meta
+	_ = k.init()
+	return nil
+}
+func (k *kubeMQEvents) setPublishStream() error {
+	var err error
+	k.publishFunc, err = k.client.Stream(k.ctx, func(err error) {
+		select {
+		case k.resultChan <- err:
+		default:
+		}
+	})
+	return err
+}
+func (k *kubeMQEvents) Publish(req *pubsub.PublishRequest) error {
+	if !k.isInitialized {
+		if err := k.init(); err != nil {
+			return err
+		}
+	}
+	k.logger.Debugf("kubemq pub/sub: publishing message to %s", req.Topic)
+	event := &kubemq.Event{
+		Id:       "",
+		Channel:  req.Topic,
+		Metadata: "",
+		Body:     req.Data,
+		ClientId: k.metadata.clientID,
+		Tags:     map[string]string{},
+	}
+	if err := k.publishFunc(event); err != nil {
+		k.logger.Errorf("kubemq pub/sub error: publishing to %s failed with error: %s", req.Topic, err.Error())
+		return err
+	}
+	return nil
+}
+func (k *kubeMQEvents) Features() []pubsub.Feature {
+	return nil
+}
+
+func (k *kubeMQEvents) Subscribe(ctx context.Context, req pubsub.SubscribeRequest, handler pubsub.Handler) error {
+	if !k.isInitialized {
+		if err := k.init(); err != nil {
+			return err
+		}
+	}
+	clientID := k.metadata.clientID
+	if clientID == "" {
+		clientID = uuid.New()
+	}
+	k.logger.Debugf("kubemq pub/sub: subscribing to %s", req.Topic)
+	err := k.client.Subscribe(ctx, &kubemq.EventsSubscription{
+		Channel:  req.Topic,
+		Group:    k.metadata.group,
+		ClientId: clientID,
+	}, func(event *kubemq.Event, err error) {
+		if err != nil {
+			k.logger.Errorf("kubemq pub/sub error: subscribing to %s failed with error: %s", req.Topic, err.Error())
+			return
+		}
+		if ctx.Err() != nil {
```

Comment:
Wouldn't this be captured by the `if err != nil` above?

Reply:
Not always, a message can be retrieved and at the same time the ctx will be cancelled.
(It happened in conformance test in some runs)

Reply:
Sounds like a bug in the upstream SDK that should handle context cancellation :) But ok, I suppose.

Reply:
yes, bad error text